### PR TITLE
implement fallback to basic authentication when GSSAPI/Negotiate fails

### PR DESCRIPTION
**Pull request by aroth-fastprotect**
Required when accessing a server which offers GSSAPI, but does not belong to the same REALM as the client.

### CODE REVIEW

**Comment by AaronOpfer** [comment]
So I applied this to my local branch and it actually fixes the issue I opened yesterday #44 . What issues are preventing its merging?

File: ngx_http_auth_spnego_module.c [hunk 1]
```diff
@@ -952,12 +987,32 @@ ngx_http_auth_spnego_handler(
...
+            } else {
+                ctx->ret = NGX_HTTP_FORBIDDEN;
+            }
+            return ctx->ret;
         }
 
         if (!ngx_spnego_authorized_principal(r, &r->headers_in.user, alcf)) {
             spnego_debug0("User not authorized");
-            return (ctx->ret = NGX_HTTP_FORBIDDEN);
+            if(alcf->allow_basic) {
```
Comment (stnoonan) [inline ngx_http_auth_spnego_module.c]: I don't think adding this here is appropriate.  If the user was successfully authenticated using the provided credentials, but isn't in the list of authorized users, a 403 is correct.  You should not fall back to basic.

File: ngx_http_auth_spnego_module.c [hunk 2]
```diff
@@ -952,12 +987,32 @@ ngx_http_auth_spnego_handler(
          * back to basic here... */
         if (NGX_DECLINED == ret) {
             spnego_debug0("GSSAPI failed");
-            return (ctx->ret = NGX_HTTP_FORBIDDEN);
+            if(alcf->allow_basic) {
```
Comment (stnoonan) [inline ngx_http_auth_spnego_module.c]: I'd prefer this be filled with return statements and the nesting reduced.

**Comment by Neptunao** [comment]
Hey guys. Any progress on it?

**Comment by Neptunao** [comment]
@stnoonan @aroth-fastprotect I created the PR based on this one with all issues fixed. https://github.com/stnoonan/spnego-http-auth-nginx-module/pull/71

**Pull request closed without merging**